### ISSUE TITLE
Possibility provide custom range for brightness

### ISSUE THREAD

**Issue by ivan-hryshko** [remcoder/gcode-preview#85]
I have a problem with displaying black (#000000), gray(#808080) and white(#FFFFFF) they all look the same , 
Is it possible to set range for:
black color for example brightness in the range [0-30]%
for gray [30-70]%
And for white [70-100]%
I think it will make the display of models more realistic
<img width="367" alt="image" src="https://github.com/remcoder/gcode-preview/assets/13918933/dcb846d2-c18c-4c69-8a9f-d0a676903ac9">

**Comment by remcoder** [comment]
Good point. It is a limition of the current implementation.

I'm trying to think of a way that works also for non black/gray/white colors. And I'm reluctant to add more config options but I'm also reluctant to change existing behavior.

Do you have some ideas?

Also to get this moving a PR and some test cases would be really helpful 😄

**Comment by ivan-hryshko** [comment]
@remcoder  i dont have permision to make PR, but i. thing about something like this
 ```
 minBrightness: number = 10;
 maxBrightness: number = 70;`
```

```
    this.maxBrightness = this.validateBrightness(opts.maxBrightness) ?? this.maxBrightness;
    this.minBrightness = this.validateBrightness(opts.maxBrightness) ?? this.minBrightness;
```
 ```
validateBrightness(value: number) {
    if (value === undefined || value === null || typeof value !== 'number') {
      return undefined
    }
    if (value > 100) {
      return 100
    }
    if (value < 0) {
      return 0
    }
    return value
  }
```

 ```
     
        const preparedMaxBrightness = this.maxBrightness / 100
        const preparedMinBrightness = this.minBrightness / 100
        const brightness = preparedMinBrightness + (preparedMaxBrightness * index) / this.layers.length;
```

**Comment by remcoder** [comment]
Thanks for your response 🙂 

It would work but it also feels a little bit ad hoc.

I want the preview to Just Work for most gcode files without much tuning. But you're right that currently the shading doesn't work well with white/gray/black.

So I'm inclined to have a built-in way (so no extra settings) of reducing the brightness range when grey colors are used (low saturation cases).

But we also need to keep in mind that this isn't true lighting so it will never look as good. For instance the more we reduce the brightness range, the harder it is so see detail. 

What are your ideas on this?

FYI, if you want to know how you can send a PF: create a fork of this repo first, commit changes to your forked repo, and then send me a PR.

**Comment by ivan-hryshko** [comment]
[https://github.com/remcoder/gcode-preview/pull/92](https://github.com/remcoder/gcode-preview/pull/92 ) 
@remcoder mode extrusion as tube look amazing it fully resolve my problem displaying black and gray colors.
When do you plan to add this mod to master? Can this mode be used and does it still need to be tested?

**Comment by remcoder** [comment]
Tube rendering is under development by @sophiedeziel ~so let's give her some time to complete it~ and we haven't exactly discussed yet at what point to release it. There is no fixed scope or schedule but you can follow the discussion in issue #89 .

You're free to use the current implementation of course but it will probably change still. For the time being I think you can reference the github repo and branch in your package.json. Please let us know of any findings when you encounter them!

**Comment by remcoder** [comment]
update: tube rendering was [released as v2.12](https://github.com/remcoder/gcode-preview/releases/tag/v2.12.0) albeit still experimental. 
 
@ivan-hryshko  Can I close this ticket?

**Comment by ivan-hryshko** [comment]
@remcoder i think yes
